File: go_engine/engine/queue.go
package engine

import (
	"time"

	"github.com/google/uuid"
)

type TaskQueue struct {
	persistence *SqlitePersistence
}

func NewTaskQueue(persistence *SqlitePersistence) *TaskQueue {
	return &TaskQueue{persistence: persistence}
}

func (q *TaskQueue) EnqueueTask(runID, taskType, queueName string, payload map[string]interface{}, scheduledAt *float64, timeoutSeconds *int, heartbeatTimeoutSeconds *int, maxAttempts int) (string, error) {
	taskID := uuid.NewString()
	now := float64(time.Now().UnixNano()) / 1e9
	scheduled := now
	if scheduledAt != nil {
		scheduled = *scheduledAt
	}
	var timeoutAt *float64
	if timeoutSeconds != nil {
		value := scheduled + float64(*timeoutSeconds)
		timeoutAt = &value
	}
	task := Task{
		TaskID:                  taskID,
		RunID:                   runID,
		TaskType:                taskType,
		QueueName:               queueName,
		Status:                  "PENDING",
		ScheduledAt:             scheduled,
		Attempt:                 0,
		MaxAttempts:             maxAttempts,
		Payload:                 payload,
		HeartbeatTimeoutSeconds: heartbeatTimeoutSeconds,
		TimeoutAt:               timeoutAt,
	}
	if err := q.persistence.InsertTask(task); err != nil {
		return "", err
	}
	return taskID, nil
}

func (q *TaskQueue) LeaseTasks(queueName string, limit int, leaseSeconds int) ([]Task, error) {
	return q.persistence.LeaseTasks(queueName, limit, leaseSeconds)
}

func (q *TaskQueue) CompleteTask(taskID string) error {
	now := float64(time.Now().UnixNano()) / 1e9
	return q.persistence.UpdateTaskStatus(taskID, "COMPLETED", &now, nil)
}

func (q *TaskQueue) FailTask(taskID string, error string) error {
	now := float64(time.Now().UnixNano()) / 1e9
	return q.persistence.UpdateTaskStatus(taskID, "FAILED", &now, &error)
}

func (q *TaskQueue) RescheduleTask(taskID string, scheduledAt float64) error {
	return q.persistence.UpdateTaskSchedule(taskID, scheduledAt)
}

func (q *TaskQueue) Heartbeat(taskID string) error {
	now := float64(time.Now().UnixNano()) / 1e9
	return q.persistence.UpdateTaskHeartbeat(taskID, now)
}

func (q *TaskQueue) GetTask(taskID string) (*Task, error) {
	return q.persistence.GetTask(taskID)
}

File: python_engine/workflow_engine/queue.py
"""
Task queue abstraction backed by SQLite.
"""

from __future__ import annotations

import time
import uuid
from typing import Any, Dict, List, Optional

from .models import Task
from .persistence import SqlitePersistence


class TaskQueue:
    def __init__(self, persistence: SqlitePersistence) -> None:
        self._persistence = persistence

    def enqueue_task(
        self,
        run_id: str,
        task_type: str,
        queue_name: str,
        payload: Dict[str, Any],
        scheduled_at: Optional[float] = None,
        timeout_seconds: Optional[int] = None,
        heartbeat_timeout_seconds: Optional[int] = None,
        max_attempts: int = 1,
    ) -> str:
        task_id = str(uuid.uuid4())
        now = time.time()
        scheduled = scheduled_at if scheduled_at is not None else now
        timeout_at = None
        if timeout_seconds is not None:
            timeout_at = scheduled + timeout_seconds
        task = Task(
            task_id=task_id,
            run_id=run_id,
            task_type=task_type,
            queue_name=queue_name,
            status="PENDING",
            scheduled_at=scheduled,
            leased_until=None,
            completed_at=None,
            attempt=0,
            max_attempts=max_attempts,
            payload=payload,
            last_heartbeat_at=None,
            heartbeat_timeout_seconds=heartbeat_timeout_seconds,
            timeout_at=timeout_at,
            last_error=None,
        )
        self._persistence.insert_task(task)
        return task_id

    def lease_tasks(self, queue_name: str, limit: int, lease_seconds: int) -> List[Task]:
        now = time.time()
        return self._persistence.lease_tasks(queue_name, now, limit, lease_seconds)

    def complete_task(self, task_id: str) -> None:
        self._persistence.update_task_status(task_id, "COMPLETED", completed_at=time.time())

    def fail_task(self, task_id: str, error: str) -> None:
        self._persistence.update_task_status(task_id, "FAILED", completed_at=time.time(), last_error=error)

    def reschedule_task(self, task_id: str, scheduled_at: float) -> None:
        self._persistence.update_task_schedule(task_id, scheduled_at)

    def heartbeat(self, task_id: str) -> None:
        self._persistence.update_task_heartbeat(task_id, time.time())

    def get_task(self, task_id: str) -> Optional[Task]:
        return self._persistence.get_task(task_id)
